### PR TITLE
Ring / Circular Buffer Library

### LINKED ISSUE #122: Circular Buffer Support
Somewhat linked to #71 since we need some sort of input buffer.

### PR DESCRIPTION
**Pull request by Kfeavel**
Closes #122

### CODE REVIEW

**Comment by Kfeavel** [comment]
I can't tell if the commit order is reversed but the resulting code seems to be fine.

File: kernel/lib/ring_buffer.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        errno = EINVAL;
+    }
+    // Return our status code
+    return status;
+}
+
+void px_ring_buffer_destroy(px_ring_buff_t* buff)
+{
+    // Free our data buffer and set the struct to null
+    free(buff->data);
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp]: We probably should check for `null` here

File: kernel/lib/ring_buffer.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    int status = 0;
+    if (buff != NULL) {
+        // Check if the buffer is empty. If so, we can't dequeue
+        if (px_ring_buffer_is_empty(buff)) {
+            status = -1;
+            errno = EINVAL;
+        } else {
+            // Read out the data and decrement the position
+            *data = buff->data[buff->tail];
+            buff->tail = ((buff->tail + 1) & (buff->size - 1));
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp]: I'm pretty sure this won't work. Maybe you meant to use `%`?

File: kernel/lib/ring_buffer.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    int status = 0;
+    if (buff != NULL) {
+        // Check if the buffer is full. If so, we can't enqueue.
+        if (px_ring_buffer_is_full(buff)) {
+            status = -1;
+            errno = ENOBUFS;
+        } else {
+            // Write the data at the write index
+            buff->data[buff->head] = byte;
+            buff->head = ((buff->head + 1) & (buff->size - 1));
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp]: See comment comment below.

**Review by micahswitzer — CHANGES_REQUESTED** [review]

**Comment by Kfeavel** [comment]
Changes made. Ready for re-review.

File: kernel/lib/ring_buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+int px_ring_buffer_destroy(px_ring_buff_t* buff) {
+    int status = 0;
+    if (buff != NULL) {
+        // Free our data buffer and set the struct to null
+        free(buff->data);
+        buff = NULL;
+    } else {
+        status = -1;
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp:49]: Set `errno` to `einval`?

File: kernel/lib/ring_buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    int status = 0;
+    if (buff != NULL) {
+        // Check if the buffer is full. If so, we can't enqueue.
+        if (px_ring_buffer_is_full(buff)) {
+            status = -1;
+            errno = ENOBUFS;
+        } else {
+            // Write the data at the write index
+            buff->data[buff->head] = byte;
+            buff->head = ((buff->head + 1) % (buff->size - 1));
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp]: Should be `buff->size` not minus one

File: kernel/lib/ring_buffer.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    int status = 0;
+    if (buff != NULL) {
+        // Check if the buffer is empty. If so, we can't dequeue
+        if (px_ring_buffer_is_empty(buff)) {
+            status = -1;
+            errno = EINVAL;
+        } else {
+            // Read out the data and decrement the position
+            *data = buff->data[buff->tail];
+            buff->tail = ((buff->tail + 1) % (buff->size - 1));
```
Comment (micahswitzer) [inline kernel/lib/ring_buffer.cpp]: See above comment

**Review by micahswitzer — APPROVED** [review]

**Pull request merged**